File: ios/MeasureSDK/Events/CustomEventCollector.swift
//
//  CustomEventCollector.swift
//  MeasureSDK
//
//  Created by Adwin Ross on 24/12/24.
//

import Foundation

protocol CustomEventCollector {
    func enable()
    func disable()
    func trackEvent(name: String, attributes: [String: AttributeValue], timestamp: Int64?)
}

final class BaseCustomEventCollector: CustomEventCollector {
    private let logger: Logger
    private let eventProcessor: EventProcessor
    private let timeProvider: TimeProvider
    private let configProvider: ConfigProvider
    private var isEnabled: Bool = false
    private lazy var customEventNameRegex: NSRegularExpression? = {
        try? NSRegularExpression(pattern: configProvider.customEventNameRegex)
    }()

    init(logger: Logger, eventProcessor: EventProcessor, timeProvider: TimeProvider, configProvider: ConfigProvider) {
        self.logger = logger
        self.eventProcessor = eventProcessor
        self.timeProvider = timeProvider
        self.configProvider = configProvider
    }

    func enable() {
        isEnabled = true
    }

    func disable() {
        isEnabled = false
    }

    func trackEvent(name: String, attributes: [String: AttributeValue], timestamp: Int64?) {
        guard isEnabled else { return }
        guard validateName(name) else { return }
        guard validateAttributes(name: name, attributes: attributes) else { return }

        let data = CustomEventData(name: name)
        let userDefinedAttributes = EventSerializer.serializeUserDefinedAttribute(attributes)

        eventProcessor.trackUserTriggered(data: data,
                                          timestamp: timestamp ?? timeProvider.now(),
                                          type: .custom,
                                          attributes: nil,
                                          sessionId: nil,
                                          attachments: nil,
                                          userDefinedAttributes: userDefinedAttributes)
    }

    private func validateName(_ name: String) -> Bool {
        if name.isEmpty {
            logger.log(level: .warning, message: "Event name is empty. This event will be dropped.", error: nil, data: nil )
            return false
        }

        if name.count > configProvider.maxEventNameLength {
            logger.log(level: .warning, message: "Event(\(name)) exceeded max allowed length. This event will be dropped.", error: nil, data: nil)
            return false
        }

        if let regex = customEventNameRegex,
           regex.firstMatch(in: name, options: [], range: NSRange(location: 0, length: name.count)) == nil {
            logger.log(level: .warning, message: "Event(\(name)) does not match the allowed pattern. This event will be dropped.", error: nil, data: nil)
            return false
        }

        return true
    }

    private func validateAttributes(name: String, attributes: [String: AttributeValue]) -> Bool {
        if attributes.count > configProvider.maxUserDefinedAttributesPerEvent {
            logger.log(level: .warning, message: "Event(\(name)) contains more than max allowed attributes. This event will be dropped.", error: nil, data: nil)
            return false
        }

        return attributes.allSatisfy { key, value in
            let isKeyValid = validateKey(key)
            let isValueValid = validateValue(value)

            if !isKeyValid {
                logger.log(level: .warning, message: "Event(\(name)) contains invalid attribute key: \(key). This event will be dropped.", error: nil, data: nil)
            }
            if !isValueValid {
                logger.log(level: .warning, message: "Event(\(name)) contains invalid attribute value. This event will be dropped.", error: nil, data: nil)
            }

            return isKeyValid && isValueValid
        }
    }

    private func validateKey(_ key: String) -> Bool {
        return key.count <= configProvider.maxUserDefinedAttributeKeyLength
    }

    private func validateValue(_ value: AttributeValue) -> Bool {
        switch value {
        case .string(let stringValue):
            return stringValue.count <= configProvider.maxUserDefinedAttributeValueLength
        default:
            return true
        }
    }
}

File: android/measure/src/main/java/sh/measure/android/events/CustomEventCollector.kt
package sh.measure.android.events

import sh.measure.android.attributes.AttributeValue
import sh.measure.android.attributes.StringAttr
import sh.measure.android.config.ConfigProvider
import sh.measure.android.logger.LogLevel
import sh.measure.android.logger.Logger
import sh.measure.android.utils.TimeProvider
import java.util.concurrent.atomic.AtomicBoolean

internal class CustomEventCollector(
    private val logger: Logger,
    private val signalProcessor: SignalProcessor,
    private val timeProvider: TimeProvider,
    private val configProvider: ConfigProvider,
) {
    private val isEnabled = AtomicBoolean(false)
    private val customEventNameRegex by lazy { Regex(configProvider.customEventNameRegex) }

    fun register() {
        isEnabled.set(true)
    }

    fun unregister() {
        isEnabled.set(false)
    }

    fun trackEvent(name: String, attributes: Map<String, AttributeValue>, timestamp: Long?) {
        if (!isEnabled.get()) {
            return
        }
        if (!validateName(name)) {
            return
        }
        if (!validateAttributes(name, attributes)) {
            return
        }
        val data = CustomEventData(name)
        signalProcessor.track(
            timestamp = timestamp ?: timeProvider.now(),
            type = EventType.CUSTOM,
            data = data,
            userTriggered = true,
            userDefinedAttributes = attributes,
        )
        logger.log(LogLevel.Debug, "Custom event($name) received")
    }

    private fun validateName(name: String): Boolean {
        if (name.isEmpty()) {
            logger.log(
                LogLevel.Error,
                "Invalid event: name is empty",
            )
            return false
        }

        if (name.length > configProvider.maxEventNameLength) {
            logger.log(
                LogLevel.Error,
                "Invalid event($name): name exceeds maximum length of ${configProvider.maxEventNameLength} characters",
            )
            return false
        }

        if (!name.matches(customEventNameRegex)) {
            return false
        }

        return true
    }

    private fun validateAttributes(name: String, attributes: Map<String, AttributeValue>): Boolean {
        if (attributes.size > configProvider.maxUserDefinedAttributesPerEvent) {
            logger.log(
                LogLevel.Error,
                "Invalid event($name): exceeds maximum of ${configProvider.maxUserDefinedAttributesPerEvent} attributes",
            )
            return false
        }

        return attributes.all { (key, value) ->
            val isKeyValid = isKeyValid(key)
            val isValueValid = isValueValid(value)
            if (!isKeyValid) {
                logger.log(
                    LogLevel.Error,
                    "Invalid event($name): invalid attribute key: $key",
                )
            }
            if (!isValueValid) {
                logger.log(
                    LogLevel.Error,
                    "Invalid event($name): invalid attribute value: $value",
                )
            }
            isKeyValid && isValueValid
        }
    }

    private fun isKeyValid(key: String): Boolean = key.length <= configProvider.maxUserDefinedAttributeKeyLength

    private fun isValueValid(value: AttributeValue): Boolean = when (value) {
        is StringAttr -> value.value.length <= configProvider.maxUserDefinedAttributeValueLength
        else -> true
    }
}
